crates/wasi-http/src/p3/request.rs

```diff
@@ -348,3 +494,126 @@ pub async fn default_send_request(
+    use std::task::{Context, Waker};
+    use wasmtime::{Engine, Store};
+    use wasmtime_wasi::{ResourceTable, WasiCtx, WasiCtxBuilder, WasiCtxView, WasiView};
+
+    struct TestHttpCtx;
+    struct TestCtx {
+        table: ResourceTable,
+        wasi: WasiCtx,
+        http: TestHttpCtx,
+    }
+
+    impl TestCtx {
+        fn new() -> Self {
+            Self {
+                table: ResourceTable::default(),
+                wasi: WasiCtxBuilder::new().build(),
+                http: TestHttpCtx,
+            }
+        }
+    }
+
+    impl WasiView for TestCtx {
+        fn ctx(&mut self) -> WasiCtxView<'_> {
+            WasiCtxView {
+                ctx: &mut self.wasi,
+                table: &mut self.table,
+            }
+        }
+    }
+
+    impl WasiHttpCtx for TestHttpCtx {}
+
+    impl WasiHttpView for TestCtx {
+        fn http(&mut self) -> WasiHttpCtxView<'_> {
+            WasiHttpCtxView {
+                ctx: &mut self.http,
+                table: &mut self.table,
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn test_request_into_http_schemes() -> Result<()> {
+        let schemes = vec![Some(Scheme::HTTP), Some(Scheme::HTTPS), None];
+        let engine = Engine::default();
+
+        for scheme in schemes {
+            let (req, fut) = Request::new(
+                Method::GET,
+                scheme.clone(),
+                Some(Authority::from_static("example.com")),
+                Some(PathAndQuery::from_static("/path?query=1")),
+                HeaderMap::new(),
+                None,
+                Full::new(Bytes::from_static(b"body"))
+                    .map_err(|x| match x {})
+                    .boxed(),
+            );
+            let mut store = Store::new(&engine, TestCtx::new());
+            let http_req = req.into_http(&mut store, async { Ok(()) }).unwrap();
+            assert_eq!(http_req.method(), Method::GET);
+            let expected_scheme = scheme.unwrap_or(Scheme::HTTPS); // default scheme
+            assert_eq!(
+                http_req.uri(),
+                &http::Uri::from_str(&format!(
+                    "{}://example.com/path?query=1",
+                    expected_scheme.as_str()
+                ))
+                .unwrap()
+            );
+            let body_bytes = http_req.into_body().collect().await?;
+            assert_eq!(*body_bytes.to_bytes(), *b"body");
+            let mut cx = Context::from_waker(Waker::noop());
+            let mut fut = Box::pin(fut);
+            let result = fut.as_mut().poll(&mut cx);
+            assert!(matches!(result, futures::task::Poll::Ready(Ok(()))));
+        }
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_request_into_http_invalid_content_length() -> Result<()> {
+        let engine = Engine::default();
+        let mut headers = HeaderMap::new();
+        headers.insert(
+            http::header::CONTENT_LENGTH,
+            HeaderValue::from_static("invalid"),
```

Comment:
I don't think this is valid. The `Request::new` assumes that the host has performed the validation if the request is of foreign origin. For example, if request was received using `hyper`, it would have to be a valid HTTP request (e.g. with compliant `content-length`)

This is not a valid `content-length` header value though, so `Request::new` should not be called with it